src/py/mat3ra/made/tools/analyze.py

```diff
@@ -481,3 +481,44 @@ def get_undercoordinated_atom_indices(
     coordination_numbers = get_coordination_numbers(material, indices, cutoff)
     undercoordinated_atoms_indices = [i for i, cn in enumerate(coordination_numbers) if cn <= coordination_threshold]
     return undercoordinated_atoms_indices
+
+
+def get_optimal_displacements(
+    material: Material,
+    grid_size: Tuple[int, int] = (10, 10),
+    search_range: Tuple[float, float] = (-1.0, 1.0),
+    shadowing_radius: float = 2.5,
+) -> List[List[float]]:
+    """
+    Return all optimal displacements for the interface material by calculating
+    the norm of distances for each film atom to substrate atoms within a certain radius.
+    The displacement is done on a grid and the displacement that yields minimum norm is returned.
+
+    Args:
+        material (Material): The interface Material object.
+        grid_size (Tuple[int, int]): The size of the grid.
+        search_range (Tuple[float, float]): The range to search for optimal displacements.
+        shadowing_radius (float): The shadowing radius to detect the surface atoms, in Angstroms.
+
+    Returns:
+        List[List[float]]: The optimal displacements.
+    """
+    from .calculate import calculate_norm_of_distances
+    from .modify import displace_interface
+
+    x_values = np.linspace(search_range[0], search_range[1], grid_size[0])
+    y_values = np.linspace(search_range[0], search_range[1], grid_size[1])
+
+    norms = np.zeros(grid_size)
+
+    for i, x in enumerate(x_values):
+        for j, y in enumerate(y_values):
+            displaced_material = displace_interface(material, [x, y, 0], use_cartesian_coordinates=True)
+            norm = calculate_norm_of_distances(displaced_material, shadowing_radius)
+            norms[i, j] = norm
```

Comment:
The above 3 lines can be separated into a function and `get_optimal_displacements` can be adjusted into a more genering `calculate_on_xy_grid`